File: templates/problems/10_backtracking/09_n_queens/typescript/solution.ts
/**
 * Solve the N-Queens problem using backtracking.
 * Time: O(n!), Space: O(n) for column and diagonal tracking
 */
export function solveSolveNQueens(n: number): string[][] {
    const result: string[][] = [];
    const cols = new Array<boolean>(n).fill(false);
    const diag1 = new Map<number, boolean>(); // row - col
    const diag2 = new Map<number, boolean>(); // row + col
    const board: string[][] = Array.from({ length: n }, () =>
        new Array<string>(n).fill('.')
    );

    const backtrack = (row: number): void => {
        if (row === n) {
            result.push(board.map((r) => r.join('')));
            return;
        }
        for (let col = 0; col < n; col++) {
            if (cols[col] || diag1.get(row - col) || diag2.get(row + col)) {
                continue;
            }
            board[row][col] = 'Q';
            cols[col] = true;
            diag1.set(row - col, true);
            diag2.set(row + col, true);
            backtrack(row + 1);
            board[row][col] = '.';
            cols[col] = false;
            diag1.set(row - col, false);
            diag2.set(row + col, false);
        }
    };

    backtrack(0);
    return result;
}

File: templates/problems/10_backtracking/09_n_queens/python/solution.py
def solve_solve_n_queens(n: int) -> list[list[str]]:
    """Solve the N-Queens problem using backtracking.

    Time: O(n!), Space: O(n) for column and diagonal tracking
    """
    result: list[list[str]] = []
    cols = [False] * n
    diag1: dict[int, bool] = {}  # row - col
    diag2: dict[int, bool] = {}  # row + col
    board = [["." for _ in range(n)] for _ in range(n)]

    def backtrack(row: int) -> None:
        if row == n:
            result.append(["".join(r) for r in board])
            return
        for col in range(n):
            if cols[col] or diag1.get(row - col, False) or diag2.get(row + col, False):
                continue
            board[row][col] = "Q"
            cols[col] = True
            diag1[row - col] = True
            diag2[row + col] = True
            backtrack(row + 1)
            board[row][col] = "."
            cols[col] = False
            diag1[row - col] = False
            diag2[row + col] = False

    backtrack(0)
    return result
